universum/configuration_support.py

```diff
@@ -1,19 +1,286 @@
+
+    .. testcode::
+        :hide:
+
+        print("$ ./configs.py")
+        print(configs.dump())
+
+    .. testoutput::
+
+        $ ./configs.py
+        [{'name': 'Make Linux', 'command': 'make --platform Linux', 'pass_tag': 'pass_Linux'},
+        {'name': 'Make Windows', 'command': 'make --platform Windows', 'pass_tag': 'pass_Windows'}]
+
+    This means that tags "pass_Linux" and "pass_Windows" will be sent to TeamCity's build.
+
+
+.. note::
+
+    All the paths, specified in `command`, `artifacts` and `directory` parameters, can be
+    absolute or relative. All relative paths start from the project root (see :ref:`get_project_root`).
 
-def combine(dictionary_a: Dict, dictionary_b: Dict) -> Dict:
+    """
+
+    # pylint: disable-msg=too-many-locals
+    def __init__(self,
+                 name: str = '',
+                 command: Optional[List[str]] = None,
+                 environment: Optional[Dict[str, str]] = None,
+                 artifacts: str = '',
+                 report_artifacts: str = '',
+                 artifact_prebuild_clean: bool = False,
+                 directory: str = '',
+                 critical: bool = False,
+                 background: bool = False,
+                 finish_background: bool = False,
+                 code_report: bool = False,
+                 pass_tag: str = '',
+                 fail_tag: str = '',
+                 if_env_set: str = '',
+                 extras: Optional[Dict[str, str]] = None,
+                 **kwargs) -> None:  # explicit constructor for client's type safety
+        self.name: str = name
+        self.command: List[str] = command if command else []
+        self.environment: Dict[str, str] = environment if environment else {}
+        self.artifacts: str = artifacts
+        self.report_artifacts: str = report_artifacts
+        self.artifact_prebuild_clean: bool = artifact_prebuild_clean
+        self.directory: str = directory
+        self.critical: bool = critical
+        self.background: bool = background
+        self.finish_background: bool = finish_background
+        self.code_report: bool = code_report
+        self.pass_tag: str = pass_tag
+        self.fail_tag: str = fail_tag
+        self.if_env_set: str = if_env_set
+        self.skip_numbering_level: bool = False
+        self.children: Optional['Variations'] = None
+        self.extras: Dict[str, str] = extras if extras else {}
+        for key, value in kwargs.items():
+            self.extras[key] = value
+
+    def replace_string(self, from_string: str, to_string: str) -> None:
+        self.command = [word.replace(from_string, to_string) for word in self.command]
+        self.artifacts = self.artifacts.replace(from_string, to_string)
+        self.report_artifacts = self.report_artifacts.replace(from_string, to_string)
+        self.directory = self.directory.replace(from_string, to_string)
+        # should we update extras? (user-defined)
```

Comment:
should we replace path in user-defined variables?

Reply:
I'd say 'yes'. As a user, I'd expect a 'variable-like' to behave as a variable always, not only in some cases.

Reply:
fixed in altest update